lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -1199,14 +1199,117 @@ struct CSEVisitor : NodeWalker {
+    // Only consider Vars for CSE if they have the same type. The train kind and
+    // visibility must already be the same.
+    if (lhs->getType() != rhs->getType()) {
+      return false;
+    }
+    assert(lhs->getVisibilityKind() == rhs->getVisibilityKind() &&
+           "Should only be comparing Variables with same VisibilityKind.");
+    assert(lhs->getTrainKind() == rhs->getTrainKind() &&
+           "Should only be comparing Variables with same TrainKind.");
+    return lhs->getPayload().isEqual(rhs->getPayload(), /* allowedError */ 0.0);
+  }
+};
+
 } // namespace
 
+/// Determine which vars are written into in \p M, either as a result, or as an
+/// overwritten input, and insert them into \p varsWithWriters. This needs to be
+/// done across all functions, as this combines variables which may be used
+/// across multiple functions.
+static void findVarsWithWriters(Module *M,
+                                std::unordered_set<Node *> *varsWithWriters) {
+  for (auto *F : M->getFunctions()) {
+    for (auto &N : F->getNodes()) {
+      for (unsigned i = 0; i < N.getNumInputs(); i++) {
+        if (N.isOverwrittenNthInput(i)) {
+          auto overwrittenInputNode = N.getNthInput(i);
+          if (isa<Variable>(overwrittenInputNode)) {
+            varsWithWriters->insert(overwrittenInputNode);
+          }
+        }
+      }
+      for (unsigned i = 0; i < N.getNumResults(); i++) {
+        auto resNode = N.getNthResult(i);
+        if (isa<Variable>(resNode)) {
+          varsWithWriters->insert(resNode);
+        }
+      }
+    }
+  }
+}
+
+/// Common Subexpression Elimination for Variables.
+static void varsCSE(Module *M) {
+  // Find all Vars that have writers.
+  std::unordered_set<Node *> varsWithWriters;
+  findVarsWithWriters(M, &varsWithWriters);
+
+  std::unordered_map<Variable *, Variable *, VarsHasherCSE, VarsEqCSE> cseVars;
+  for (auto &V : M->getVars()) {
+    // Perform CSE on all vars that have no writers.
+    if (varsWithWriters.find(V) != varsWithWriters.end()) {
+      continue;
+    }
+
+    // Only look for var CSE on vars of small enough size. Otherwise just skip
+    // them. 256 is used as a heuristic, to keep compile time reasonable.
+    constexpr size_t maxNumEls = 256;
+    size_t numEls = V->getType()->size();
+    if (numEls > maxNumEls) {
+      continue;
+    }
+
+    // Only perform CSE on private variables that have no train kind.
+    if (V->getVisibilityKind() != VisibilityKind::Private ||
+        V->getTrainKind() != Variable::TrainKind::None) {
+      continue;
+    }
+
+    // Try to find a var that has the same data as the current one.
+    auto foundI = cseVars.find(V);
+    if (foundI == cseVars.end()) {
+      // No node CSE-equivalent to the current one has been seen yet. Remember
+      // this variable, so that the next occurrence can be replaced by this
+      // one.
+      cseVars.insert({V, V});
+      assert(cseVars.find(V) != cseVars.end());
```

Comment:
why is this needed? (we just inserted V into the map)

Reply:
@rdzhabarov I think it is mostly copy/pasted from the CSE pass that I've written a while ago. IIRC, the initial idea of doing such a defensive check was to make sure that our custom hash and equality functions work properly when the insertion is performed. There were some issues with that at the beginning.

Reply:
Yep, what @opti-mix said. I can remove it if desired, but I did think it was useful.